node/node-sdk.md

```diff
@@ -0,0 +1,187 @@
+# NodeJS SDK Proposal
+
+NodeJS is a perfect candidate for authoring Temporal workflows, by leveraging V8 isolates for running workflow code,
+we can create a JS runtime that's completely deterministic.
+
+The SDK steers developers to write their workflows and activities in TypeScript but vanilla JS is also supported.
+All examples in this proposal are written in TypeScript.
+
+## Table Of Contents
+- [Activities](#activities)
+    - [Examples](#examples)
+    - [Registration](#registration)
+    - [Payloads](#payloads)
+- [Workflows](#workflows)
+    - [Example](#example)
+    - [Activity Options](#activity-options)
+    - [Activity Type Declarations](#activity-type-declarations)
+    - [Registration](#registration-1)
+    - [Limitations](#limitations)
+    - [Signals](#signals)
+    - [Queries](#queries)
+    - [Logging](#logging)
+    - [Sessions](#sessions)
+    - [Deterministic Time](#deterministic-time)
+    - [Error Handling](#error-handling)
+    - [Cancellation](#cancellation)
+    - [Long Histories](#long-histories)
+
+
+## Activities
+Activity implementations are just JS functions, they run on the default NodeJS runtime, along with the Temporal SDK, and can import arbitrary npm packages.
+Activity functions may optionally accept an `ActivityContext` as the first parameter.
+
+### Examples
+#### Synchronous function
+```typescript
+import crypto from 'crypto';
+
+export function sha256(txt: string) {
+  return crypto.createHash('sha256').update(txt).digest('hex');
+}
+```
+
+#### Asynchronous function
+```typescript
+import axios from 'axios';
+
+export async function httpGet(url: string) {
+  const response = await axios.get(url);
+  return response.data;
+}
+```
+
+#### With `ActivityContext`
+```typescript
+import { ActivityContext } from 'temporal-sdk';
+import axios from 'axios';
+
+export async function httpGet(ctx: ActivityContext, url: string) {
+  const response = await axios.get(url, {
+    onDownloadProgress: (progressEvent) => {
+      const total = parseFloat(progressEvent.currentTarget.responseHeaders['Content-Length']);
+      const current = progressEvent.currentTarget.response.length;
+      const progress = Math.floor(current / total * 100);
+      ctx.heartbeat(progress);
+    },
+  });
+  return response.data;
+}
+```
+
+### Registration
+Activities should be registered in their corresponding workers.
+
+```typescript
+worker.registerActivity('httpGet', httpGet, { passContext: true });
+```
+
+### Payloads
+Initially we'll support any JSON serializable object as payload, binary data support can be added if required.
+
+TBD: consider supporting [typed arrays][typed-arrays].
+
+## Workflows
+Worflow code looks like vanilla TypeScript.
+Workflows run in the same NodeJS process as the Temporal SDK and the activities.
+Workflow code must export a `main` function which will be used as the handler method for workflow invocations.
```

Comment:
Does it mean that we will not be able to define a workflow interface the way it is done in Java SDK?

Reply:
We can still define the interface's functions and let the workflow author decide which methods to implement.
I am missing some information on the workflow interface in Java, I'll read more about it.
We can switch to an interface if we find it more suitable.